### ISSUE TITLE
Layered Navigation shows no products when apply filters with configurable attribute and non-configurable

### ISSUE THREAD

**Issue by tangar76** [magento/magento2#12240]
### Preconditions
 1. Magento 2.2.1
 2. mysql  Ver 14.14 Distrib 5.7.20, for Linux (x86_64)
 3. PHP 7.0.25-1+ubuntu16.04.1+deb.sury.org+1
 4. Server version: Apache/2.4.18 (Ubuntu)

### Steps to reproduce

 1.  Find category with layered navigation and configurable products. In case of Magento Sample data it can be Women - Tops.
 2. Apply filter for non-configurable attribute (for example 'style general' - 'jacket' ) - https://prnt.sc/haasfk
 3. Apply filter for configurable attribute (from screenshot above 'color' - 'blue')

### Expected result
1. Get at least three products "Olivia 1/4 Zip Light Jacket", "Juno Jacket" and "Neve Studio Dance Jacket"

### Actual result

1. Message "We can't find products matching the selection." Check screenshot -- https://prnt.sc/haawr6

### Additional info
Layered Navigation can not find product **only when** customer use mix of **configurable and non-configurable attributes to filter** products. In other cases everything works as it should.
Same issue exists on Magento 2.2.0 ~~and Magento 2.1.x~~

**Comment by tangar76** [comment]
Sorry, we've just tested one more time and there are no such issue on Magento 2.1.x. Problem occurs only on Magento 2.2.x

**Comment by magento-engcom-team** [comment]
@tangar76, thank you for your report.
We've created internal ticket(s) MAGETWO-83827 to track progress on the issue.

**Comment by solwininfotech** [comment]
"I am working on this at #dmcdindia”

**Comment by magento-engcom-team** [comment]
@solwininfotech thank you for joining. Please accept team invitation [here](https://github.com/orgs/magento/invitation) and self-assign the issue.

**Comment by tangar76** [comment]
~~Seems like issue is solved in Magento 2.2.6~~
Update: issue still exists.

**Comment by singh190** [comment]
I have installed fresh Magento 2.2.6, but the issue is still there.

**Comment by magento-engcom-team** [comment]
Hi @novakivskiy. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by wontroba666** [comment]
same here.  any progress ?

**Comment by magento-engcom-team** [comment]
Hi @XxXgeoXxX. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by XxXgeoXxX** [comment]
@tangar76
Looks like a problem in method getJoinConditions in Magento 2.2. In Magento 2.3-developer branch it was fixed in commit f710f9b0a8d24ae89ae5eee97652087872503763 and all works correctly.
I had the similar problem and you can use next patch for fix your problem. I hope it helps you.
Or remove string  //sprintf('`%s`.`source_id` = `%s`.`source_id`', $mainTable, $joinTable) in  vendor/magento/module-catalog-search/Model/Search/FilterMapper/CustomAttributeFilter.php:141

diff --git vendor/magento/module-catalog-search/Model/Search/FilterMapper/CustomAttributeFilter.php vendor/magento/module-catalog-search/Model/Search/FilterMapper/CustomAttributeFilter.php
index fc93b86f5da..0b2a7949ab2 100644
--- vendor/magento/module-catalog-search/Model/Search/FilterMapper/CustomAttributeFilter.php
+++ vendor/magento/module-catalog-search/Model/Search/FilterMapper/CustomAttributeFilter.php
@@ -141,7 +141,7 @@ class CustomAttributeFilter
     {
         return [
             sprintf('`%s`.`entity_id` = `%s`.`entity_id`', $mainTable, $joinTable),
 -            sprintf('`%s`.`source_id` = `%s`.`source_id`', $mainTable, $joinTable),
+            //sprintf('`%s`.`source_id` = `%s`.`source_id`', $mainTable, $joinTable),
             $this->conditionManager->generateCondition(
                 sprintf('%s.attribute_id', $joinTable),
                 '=',

**Comment by tangar76** [comment]
@XxXgeoXxX 
Thanks for help.
As for commit you mentioned I can't see how it solves the issue. There are changes in some test xml files.
But as for fix you offered it works like a charm. Thanks.
(Magento 2.2.7)

```diff
--- .../app/code/Magento/CatalogSearch/Model/Search/FilterMapper/CustomAttributeFilter.php
+++ .../app/code/Magento/CatalogSearch/Model/Search/FilterMapper/CustomAttributeFilter.php (Unsaved)
@@ -141,7 +141,7 @@
     {
         return [
             sprintf('`%s`.`entity_id` = `%s`.`entity_id`', $mainTable, $joinTable),
-            sprintf('`%s`.`source_id` = `%s`.`source_id`', $mainTable, $joinTable),
+            // sprintf('`%s`.`source_id` = `%s`.`source_id`', $mainTable, $joinTable),
             $this->conditionManager->generateCondition(
                 sprintf('%s.attribute_id', $joinTable),
                 '=',

```

PS: never the less it is better to have this issue fixed in one of next releases.

**Comment by sdzhepa** [comment]
Hello @tangar76 and @XxXgeoXxX 

As was mentioned in the comment https://github.com/magento/magento2/pull/20358#issuecomment-457197218 the issue was fixed

**Comment by rickschippers** [comment]
This fix that is in magento 2.2.8 may fix the issue at hand here, but it slows down the layered navigation extremely in quite a few situations. For example on the dataset I have in one of my stores I select a couple of attributes in the layered navigation. The results would be near instant in 2.2.7, now with this patch applied the results take 30+ seconds to show up.

I logged the queries that 2.2.7 and 2.2.8 made. The only difference is the join condition mentioned here is removed.


**example FROM 2.2.7  (this is what is inserting in the temp table)**
>`SELECT `main_select`.`entity_id`, SUM(score) AS `relevance`
FROM (SELECT DISTINCT `search_index`.`entity_id`, (((0) + (0) + (0) + (0)) * 1) AS `score`
      FROM `catalog_product_index_eav` AS `search_index`
               INNER JOIN `catalog_product_index_eav` AS `style_group_filter`
                          ON `search_index`.`entity_id` = `style_group_filter`.`entity_id` 
			  AND `search_index`.`source_id` = `style_group_filter`.`source_id` 
			  AND `style_group_filter`.`attribute_id` = 228 
			  AND `style_group_filter`.`store_id` = 1
               INNER JOIN `catalog_product_index_eav` AS `quality_filter`
                          ON `search_index`.`entity_id` = `quality_filter`.`entity_id` 
			  AND `search_index`.`source_id` = `quality_filter`.`source_id` 
			  AND `quality_filter`.`attribute_id` = 234 
			  AND `quality_filter`.`store_id` = 1
               INNER JOIN `cataloginventory_stock_status` AS `stock_index`
                          ON stock_index.product_id = search_index.entity_id AND `stock_index`.`website_id` = 0 AND `stock_index`.`stock_status` = 1 AND
                             `stock_index`.`stock_id` = 1
               INNER JOIN `cataloginventory_stock_status` AS `sub_products_stock_index`
                          ON sub_products_stock_index.product_id = search_index.source_id AND `sub_products_stock_index`.`website_id` = 0 AND
                             `sub_products_stock_index`.`stock_status` = 1 AND `sub_products_stock_index`.`stock_id` = 1
               INNER JOIN `catalog_product_index_eav` AS `visibility_filter`
                          ON search_index.entity_id = visibility_filter.entity_id AND `visibility_filter`.`attribute_id` = 96 AND
                             `visibility_filter`.`value` in ('2', '4') AND `visibility_filter`.`store_id` = '1'
               INNER JOIN `catalog_category_product_index_store1` AS `category_ids_index`
                          ON search_index.entity_id = category_ids_index.product_id AND category_ids_index.store_id = '1'
      WHERE (search_index.store_id = '1')
        AND (category_ids_index.category_id = 150)
        AND (style_group_filter.value IN ('331', '5130', '539', '938'))
        AND (quality_filter.value IN ('12055', '610', '337'))) AS `main_select`
GROUP BY `entity_id`
ORDER BY `relevance` DESC, `entity_id` DESC
LIMIT 100000

**And the same page in 2.2.8** 
>SELECT `main_select`.`entity_id`, SUM(score) AS `relevance`
FROM (SELECT DISTINCT `search_index`.`entity_id`, (((0) + (0) + (0) + (0)) * 1) AS `score`
      FROM `catalog_product_index_eav` AS `search_index`
               INNER JOIN `catalog_product_index_eav` AS `style_group_filter`
                          **ON `search_index`.`entity_id` = `style_group_filter`.`entity_id`** 
			  AND `style_group_filter`.`attribute_id` = 228 
			  AND `style_group_filter`.`store_id` = 1
               INNER JOIN `catalog_product_index_eav` AS `quality_filter`
                          **ON `search_index`.`entity_id` = `quality_filter`.`entity_id`**
			  AND `quality_filter`.`attribute_id` = 234 
			  AND `quality_filter`.`store_id` = 1
               INNER JOIN `cataloginventory_stock_status` AS `stock_index`
                          ON stock_index.product_id = search_index.entity_id AND `stock_index`.`website_id` = 0 AND `stock_index`.`stock_status` = 1 AND
                             `stock_index`.`stock_id` = 1
               INNER JOIN `cataloginventory_stock_status` AS `sub_products_stock_index`
                          ON sub_products_stock_index.product_id = search_index.source_id AND `sub_products_stock_index`.`website_id` = 0 AND
                             `sub_products_stock_index`.`stock_status` = 1 AND `sub_products_stock_index`.`stock_id` = 1
               INNER JOIN `catalog_product_index_eav` AS `visibility_filter`
                          ON search_index.entity_id = visibility_filter.entity_id AND `visibility_filter`.`attribute_id` = 96 AND
                             `visibility_filter`.`value` in ('2', '4') AND `visibility_filter`.`store_id` = '1'
               INNER JOIN `catalog_category_product_index_store1` AS `category_ids_index`
                          ON search_index.entity_id = category_ids_index.product_id AND category_ids_index.store_id = '1'
      WHERE (search_index.store_id = '1')
        AND (category_ids_index.category_id = 150)
        AND (style_group_filter.value IN ('331', '5130', '539', '938'))
        AND (quality_filter.value IN ('12055', '610', '337'))) AS `main_select`
GROUP BY `entity_id`
ORDER BY `relevance` DESC, `entity_id` DESC
LIMIT 100000

The only changes are the removal of: `ON search_index.entity_id = style_group_filter.entity_id` in the joins.

When I count the rows being returned from both queries before the distinct is being applied. With my dataset/selections it is about 36k in 2.2.7 and about 47 million in 2.2.8. So no wonder it's being ridiculously slower in 2.2.8.  The end result after the distinct is applied is identical in this case.

It's gotten so bad, that the load on the category pages is completely overloading my server with 2.2.8, while 2.2.7 is running absolutely fine.
